portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/calendar/RecurrenceRule.java

```diff
@@ -0,0 +1,338 @@
+  }
+
+  enum ByRule {
+    BYSECOND,
+    BYMINUTE,
+    BYHOUR,
+    BYDAY,
+    BYMONTHDAY,
+    BYYEARDAY,
+    BYWEEKNO,
+    BYMONTH,
+    BYSETPOS
+  }
+
+  static List<String> byRuleList = Arrays.stream(ByRule.values()).map(a -> a.toString())
+      .collect(Collectors.toList());
+
+  enum Day {
+    SU,
+    MO,
+    TU,
+    WE,
+    TH,
+    FR,
+    SA
+  }
+
+  static List<String> dayList = Arrays.stream(Day.values()).map(a -> a.toString())
+      .collect(Collectors.toList());
+
+  enum TimeType {
+    DATETIME,
+    DATE,
+    PERIOD,
+    UNDEFINED;
+  }
+
+  /**
+   * A method for turning an RRule string into an {@link RRule}.
+   *
+   * @param rRuleString a string containing RRule information, excluding any prefixes (e.g.,
+   * RRULE:)
+   * @return an {@link RRule}
+   */
+  public static RRule parseRRuleString(String rRuleString) {
+    RRule.Builder builder = RRule.builder();
+
+    List<String> components = Arrays.asList(rRuleString.split(";"));
+    Map<ByRule, String> byRuleMapInput = new HashMap<>();
+    for (String property : components) {
+      String[] split = property.split("=");
+      String key = split[0];
+      String value = split[1];
+      if (key.equals(FREQ)) {
+        builder.freq(Freq.valueOf(value));
+      } else if (key.equals(UNTIL)) {
+        builder.until(value); // TODO(olsona): consider parsing further?
+      } else if (key.equals(COUNT)) {
+        builder.count(Integer.parseInt(value));
+      } else if (key.equals(INTERVAL)) {
+        builder.interval(Integer.parseInt(value));
+      } else if (byRuleList.contains(key)) {
+        byRuleMapInput.put(ByRule.valueOf(key), value);
+      } else if (key.equals(WKST)) {
+        builder.wkst(Day.valueOf(value));
+      }
+    }
+    if (!byRuleMapInput.isEmpty()) {
+      builder.byRuleMap(byRuleMapInput);
+    }
+    return builder.build();
+  }
+
+  /**
+   * A method for parsing an RDate string into an {@link RDate}
+   *
+   * @param rDateString a string containing RDate information, excluding any prefixes (e.g.,
+   * RDATE:)
+   * @return an {@link RDate}
+   */
+  public static RDate parseRDateString(String rDateString) {
+    RDate.Builder builder = RDate.builder();
+    StringTokenizer stringTokenizer = new StringTokenizer(rDateString, ":");
+    String token = stringTokenizer.nextToken();
+    while (stringTokenizer.hasMoreTokens()) {
+      String[] split = token.split("=");
+      String key = split[0];
+      String value = split[1];
+      if (key.equals(VALUE)) {
+        builder.value(TimeType.valueOf(value));
+      } else if (key.equals(TZID)) {
+        builder.tzidparam(value);
+      }
+      token = stringTokenizer.nextToken();
+    }
+    builder.rdtval(Arrays.asList(token.split(",")));
+
+    return builder.build();
+  }
+
+  public static ExDate parseExDateString(String exDateString) {
+    ExDate.Builder builder = ExDate.builder();
+    StringTokenizer stringTokenizer = new StringTokenizer(exDateString, ":");
```

Comment:
Do we have to check if the string is null or empty here or before this is called?